### PR TITLE
Bug 54134

### LINKED ISSUE #54134: BUG: `copy.deepcopy()` doesn't deepcopy the metadata in `.attrs`
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
import copy

# Create a GeoDataFrame
df = pd.DataFrame({
    'value': [1, 2]
})

# Add an attribute to the DataFrame
df.attrs['transform'] = {}

# Copy the DataFrame
df2 = copy.deepcopy(df)
# gdf2 = gdf.copy(deep=True)

# Print the IDs
print(id(df.attrs['transform']))  # ID of copied attribute
print(id(df2.attrs['transform']))  # ID of original attribute

# Assertion
assert df.attrs['transform'] is not df2.attrs['transform']
```

### Issue Description

I would expect the dict in .attrs to be a copy since I performed a deep copy, but it is still the same object. Originally I though it was a bug of `geopandas`, but it is a bug of `pandas`. For cross-ref, this is my original issue: https://github.com/geopandas/geopandas/issues/2920#issuecomment-1594477028

### Expected Behavior

The assertion should not fail.

### Installed Versions

<details>

>>> pd.show_versions()
/Users/macbook/miniconda3/envs/ome/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")

INSTALLED VERSIONS
------------------
commit           : 2e218d10984e9919f0296931d92ea851c6a6faf5
python           : 3.10.12.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 22.5.0
Version          : Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000
machine          : arm64
processor        : arm
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.5.3
numpy            : 1.23.4
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 67.8.0
pip              : 23.1.2
Cython           : None
pytest           : 7.4.0
hypothesis       : None
sphinx           : 4.5.0
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.14.0
pandas_datareader: None
bs4              : 4.12.2
bottleneck       : None
brotli           : 1.0.9
fastparquet      : None
fsspec           : 2023.6.0
gcsfs            : None
matplotlib       : 3.7.2
numba            : 0.57.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 12.0.1
pyreadstat       : None
pyxlsb           : None
s3fs             : 2023.6.0
scipy            : 1.11.1
snappy           : None
sqlalchemy       : 2.0.18
tables           : None
tabulate         : 0.9.0
xarray           : 2022.12.0
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : 2023.3

</details>

### PR DESCRIPTION
**Pull request by Charlie-83**
- [X] closes #54134
- [X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [X] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [X] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v2.1.0.rst [hunk 1]
```diff
@@ -615,6 +615,7 @@ Other
 - Bug in :meth:`DataFrame.reindex` with a ``fill_value`` that should be inferred with a :class:`ExtensionDtype` incorrectly inferring ``object`` dtype (:issue:`52586`)
 - Bug in :meth:`DataFrame.shift` and :meth:`Series.shift` when passing both "freq" and "fill_value" silently ignoring "fill_value" instead of raising ``ValueError`` (:issue:`53832`)
 - Bug in :meth:`DataFrame.shift` with ``axis=1`` on a :class:`DataFrame` with a single :class:`ExtensionDtype` column giving incorrect results (:issue:`53832`)
+- Bug in :meth:`NDFrame.__finalize__` where attributes wouldn't be copied from ``other`` (:issue:`54134`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:618]:
```suggestion
- Bug where a deepcopy of a :class:`DataFrame` would not copy the `attrs` (:issue:`54134`)
```

**Comment by KangarooChief** [comment]
I have no idea why that that completely unrelated test that's marked as expected to fail is failing the checks.

**Comment by mroeschke** [comment]
Could you merge in main once more?

File: pandas/core/generic.py [hunk 2]
```diff
@@ -6153,7 +6154,7 @@ def __finalize__(self, other, method: str | None = None, **kwargs) -> Self:
         """
         if isinstance(other, NDFrame):
             for name in other.attrs:
-                self.attrs[name] = other.attrs[name]
+                self.attrs[name] = deepcopy(other.attrs[name])
```
Comment (timhoffm) [inline pandas/core/generic.py:6157]: Do we want to deepcopy unconditionally. `copy.deepcopy(df)` goes to `df.copy(deep=True)` (`deep=True` also being the default). By this change, `df.copy(deep=False)` would also deepcopy. If we don't want this, we'd have to pass the copy mode from `df.copy()` down to `__finalize__`. - One way to do this is by passing a specialized *method* parameter.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**